src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,513 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+			GDB_REGNO_S1, riscv_xlen(target),
 			AC_ACCESS_REGISTER_TRANSFER | AC_ACCESS_REGISTER_POSTEXEC);
-	if (execute_abstract_command(target, command) != ERROR_OK)
+	if (execute_abstract_command(target, startup_command) != ERROR_OK)
 		return ERROR_FAIL;
 
-	/* First read has just triggered. Result is in s1. */
-	if (count == 1) {
-		uint64_t value;
-		if (register_read_direct(target, &value, GDB_REGNO_S1) != ERROR_OK)
-			return ERROR_FAIL;
-		buf_set_u64(buffer, 0, 8 * size, value);
-		log_memory_access64(address, value, size, true);
-		return ERROR_OK;
-	}
-
+	/* First read has just triggered. Result is in s1.
+	 * dm_data registers contain the previous value of s1 (garbage).
+	 */
 	if (dmi_write(target, DM_ABSTRACTAUTO,
-			1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET) != ERROR_OK)
-		goto error;
-	/* Read garbage from dmi_data0, which triggers another execution of the
-	 * program. Now dmi_data0 contains the first good result, and s1 the next
-	 * memory value. */
-	if (dmi_read_exec(target, NULL, DM_DATA0) != ERROR_OK)
-		goto error;
-
-	/* read_addr is the next address that the hart will read from, which is the
-	 * value in s0. */
-	unsigned index = 2;
-	while (index < count) {
-		riscv_addr_t read_addr = address + index * increment;
-		LOG_DEBUG("i=%d, count=%d, read_addr=0x%" PRIx64, index, count, read_addr);
-		/* The pipeline looks like this:
-		 * memory -> s1 -> dm_data0 -> debugger
-		 * Right now:
-		 * s0 contains read_addr
-		 * s1 contains mem[read_addr-size]
-		 * dm_data0 contains[read_addr-size*2]
-		 */
+				set_field(0, DM_ABSTRACTAUTO_AUTOEXECDATA, 1)) != ERROR_OK)
+		return ERROR_FAIL;
 
-		struct riscv_batch *batch = riscv_batch_alloc(target, RISCV_BATCH_ALLOC_SIZE,
-				info->dmi_busy_delay + info->ac_busy_delay);
-		if (!batch)
-			return ERROR_FAIL;
+	/* Read garbage from dm_data0, which triggers another execution of the
+	 * program. Now dm_data contains the first good result (from s1),
+	 * and s1 the next memory value.
+	 */
+	if (dmi_read_exec(target, NULL, DM_DATA0) != ERROR_OK)
+		goto clear_abstractauto_and_fail;
 
-		unsigned reads = 0;
-		for (unsigned j = index; j < count; j++) {
-			if (size > 4)
-				riscv_batch_add_dmi_read(batch, DM_DATA1);
-			riscv_batch_add_dmi_read(batch, DM_DATA0);
+	uint32_t abstractcs;
 
-			reads++;
-			if (riscv_batch_full(batch))
-				break;
-		}
+	if (wait_for_idle(target, &abstractcs) != ERROR_OK)
+		goto clear_abstractauto_and_fail;
 
-		batch_run(target, batch);
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		return ERROR_OK;
+	case CMDERR_BUSY:
+		LOG_TARGET_ERROR(target, "Unexpected busy error. This is probably a hardware bug.");
+		/* fall through */
+	default:
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+	}
+clear_abstractauto_and_fail:
+	dmi_write(target, DM_ABSTRACTAUTO, 0);
+	return ERROR_FAIL;
```

Comment:
As a defensive coding measure, and to clearly show what should happen, dere I'd recommend to not leave the "default" case un-terminated.

```suggestion
		riscv013_clear_abstract_error(target);
		goto clear_abstractauto_and_fail;
	}

clear_abstractauto_and_fail:
	dmi_write(target, DM_ABSTRACTAUTO, 0);
	return ERROR_FAIL;
```

Reply:
Addressed.